### ISSUE TITLE
tank update procedure

### ISSUE THREAD

**Issue by Dieterbe** [raintank/raintank-metric#56]
we have basically 2 options IMHO:

1) stop tank, saving chunks to disk, update tank, start it, which means it has to decode from disk, and work through a backlog of data. this is AFAIK why we did #40
3 issues with this approach:
a) the ordering problem with nsq causes data loss and gaps in data (messages from during downtime will come out of order and we have to drop metrics if they're older than previous)
b) if the upgrade comes with changes to data structures, it will be a bunch of extra work to accommodate migrating the old data into a new format (if possible), and possibly having to throw away the old chunks, creating an even bigger gap of data.
c) the datasource as a whole is effectively offline during this process.

2) launch a new tank instance parallel to the old one on a different ip/port/machine. 
the downsides are:
a) it needs to run for a while and cover all data range between now until when we have data in cassandra, ideally up to numChunks*numspan and/or whatever range our dashboards are set to so cassandra doesn't see an increased read load when we switch.
b) it's operationally a bit more complex
c) we need to implement some additional logic to not have both instances be saving the same data to cassandra, i guess some kind of signal for an instance to tell it not to save anything, and then another one to tell it to start saving starting with data from a given boundary timestamp. haven't thought too much about this.

but the upside is the three problems from 1 don't apply here, so we can do the upgrade seamlessly without *any* data loss.

There's some hope for 1a though:
* I'm hopeful that #48 + probably some parrallelized encoding/decoding will make things fast and the time gap between stop and start small, so that we're in fairly good shape until we fix ordering properly
* @woodsaj was going to experiment with max-in-flight and concurency settings, which I think will result in mostly-correct-ordering out of NSQ. but only one way to find out...

though due do 1b and 1c i think perhaps we *shouldn't* be spending too much time on save/restore of chunks to/from disk, and just go for 2 ? I think its downsides may be easy to overcome.
@woodsaj thoughts?

**Comment by woodsaj** [comment]
I do like the idea of "2", and building on this idea i think we can solve a number of problems.

at present we have 2 channels (copies of metric data). 1 for metric-tank and 1 for nsq-metrics-to-es.  Lets add an additional 2 channels so that we have 3 channels for metric-tank (1 for each availability zone).  We can then spin up an additional metric-tank in each zone.

So we still need a way to have only 1 node as the "primary" node which writes data to cassandra. Other nodes should just keep data in memory and provide additional capacity for graphite queries.

A clustering solution would be ideal for handling the assignment of the "primary" role, but we can also just handle it manually for now.  What we need to address is when the primary role moves to a new node, how do we know which chunks have not yet been committed to cassandra?

When performing upgrades, we just need to perform them 1 node at a time.

**Comment by Dieterbe** [comment]
> What we need to address is when the primary role moves to a new node, how do we know which chunks have not yet been committed to cassandra?

how about: when we signal a secondary instance that it should become primary, it can iterate its list of metrics, and for each one query cassandra what's the latest ts? we can then mark all chunks up until that one as saved.  for simplicity we can assume we won't do this right when the month changes so that it's easy to get the proper row_key. doing this on a metric by metric basis might be slow though, maybe we can do the query for many at the same time? we could also have the old primary write out a file of `metric-key t0-of-last-saved-chunk` mappings, and load in that file when promoting the new one.

**Comment by Dieterbe** [comment]
from hangout:
* cassandra -> too inefficient, have to query all rows individual
* file save / file load -> only works for manual maintenance, if primary crashes, you don't have the information.

AJ suggested commitlog idea: after every successful cassandra chunk save, primary writes a log to a nsq/kafka topic. you need at least one secondary instance that has been consuming from this for a while (>1 chunkspan) to have built-up the required state, which it can then use to know from which point on the chunks require saving. he will work on this.

**Comment by woodsaj** [comment]
just had a great call with dieter.  The approach we agreed on was.

* continuously run 2 or more metric-tank instances with only 1 acting as *primary* and writing to cassandra (the additional instances will be used for handling graphite queries.)
* The primary nodes pushes updates to a NSQ stream when ever a chunk is persisted to cassandra.  The other nodes read this stream and use it to update their own chunk save states.
* After an outage or during maintenance, a new instance becomes the *primary* and can just persist new chunks as needed.  The GC() task will handle persisting any chunks that were skipped.

main issue to consider is that the first chunk of a fresh instance will only be a partial chunk. So the primary role should always be assigned to the instance that has been running the longest.

**Comment by Dieterbe** [comment]
@woodsaj any news on this?

**Comment by Dieterbe** [comment]
@woodsaj how is this going, or do you want me to start working on this?

**Comment by woodsaj** [comment]
This is mostly done. Was just waiting on PR#67 to be merged so i could incorporate those changes.

**Comment by woodsaj** [comment]
https://github.com/raintank/raintank-metric/pull/74

**Comment by Dieterbe** [comment]
we decided on 2 so this can be closed.
